x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/GroupByOptimizer.java

```diff
@@ -0,0 +1,103 @@
+
+        private final Entry<String, SingleGroupSource> entry;
+        private final int priority;
+
+        UnmodifiableEntryWithPriority(Entry<String, SingleGroupSource> entry, int priority) {
+            this.entry = entry;
+            this.priority = priority;
+        }
+
+        public int compareTo(UnmodifiableEntryWithPriority o) {
+            return Integer.compare(priority, o.priority);
+        }
+
+        @Override
+        public String getKey() {
+            return entry.getKey();
+        }
+
+        @Override
+        public SingleGroupSource getValue() {
+            return entry.getValue();
+        }
+
+        @Override
+        public SingleGroupSource setValue(SingleGroupSource value) {
+            // not modifiable
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    private GroupByOptimizer() {}
+
+    /**
+     * Returns an ordered collection of group by fields in order to get better performance.
+     *
+     * The decision is based on the type and whether the input field is a indexed/runtime/script field
+     *
+     * TODO: take index sorting into account
+     *
+     * @param groups group by as defined by the user
+     * @param runtimeFields set of strings which fields are runtime fields
+     * @return collection in order of priority
+     */
+    static Collection<Entry<String, SingleGroupSource>> reorderGroups(Map<String, SingleGroupSource> groups, Set<String> runtimeFields) {
+        if (groups.size() == 1) {
+            return groups.entrySet();
+        }
+
+        // Arrays.sort provides stable sort (to respect the input order if priorities match), Collections.sort not
+        // as generic arrays and generic types can't be combined, the "struct" is necessary
+        UnmodifiableEntryWithPriority[] prioritizedGroups = new UnmodifiableEntryWithPriority[groups.size()];
+
+        int index = 0;
+        for (Entry<String, SingleGroupSource> groupBy : groups.entrySet()) {
+            // prefer indexed fields over runtime fields over scripts
+            int priority = groupBy.getValue().getScriptConfig() == null
+                ? runtimeFields.contains(groupBy.getValue().getField()) ? 25 : 50
+                : 0;
+
+            switch (groupBy.getValue().getType()) {
+                case DATE_HISTOGRAM:
+                    priority += 400;
+                    break;
+                case HISTOGRAM:
+                    priority += 300;
+                    break;
+                case TERMS:
+                    priority += 200;
+                    break;
+                case GEOTILE_GRID:
+                    priority += 100;
+                    break;
+                default:
+                    assert false : "new group source type misses priority definition";
+            }
+
+            prioritizedGroups[index++] = new UnmodifiableEntryWithPriority(groupBy, priority);
+        }
+
+        Arrays.sort(prioritizedGroups, (a, b) -> b.compareTo(a));
```

Comment:
Could `Collections.reverseOrder()` be used as a comparator here?

Reply:
I am not reversing the order, I want to re-order according to the priority I calculated from inspecting the group by definition. In case the priority equals the order of the config should remain that's why I need a stable sort, it respects the insertion order.

`Collections.sort` does not provide stable sort. But re-thinking about it, I could solve that by boosting the 1st group with `groups.size()`, the 2nd with `groups.size() - 1` and so on.